### PR TITLE
Java object factory: initialize AbstractStringBuilder-derived types correctly

### PR DESCRIPTION
**Pull request by smowton**
These are currently initialized as if they are directly derived from Object, which causes a crash due to the
type inconsistency between StringBuilder { AbstractStringBuilder { Object { } } } and StringBuilder { Object { } }

With this change they are initialised more like normal types, which has the side-effect that any fields they possess that are *not* special-cased by `initialize_nondet_string_fields` are nondet-initialized as for any other type.

I add a test verifying this new behaviour, and a simpler test checking that Builder and Buffer are usable as nondet arguments at all.

***Note to reviewers: this is way easier to review with whitespace changes ignored***

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 643e9f10b6781939ac18eb43c599c8e5b8798134).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77258251
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

**Comment by thk123** [comment]
Linking my PR as related to the same type, but I think these are fundamentally different problems. #2479

**Comment by smowton** [comment]
Yeah that's different

**Comment by smowton** [comment]
TG bump: https://github.com/diffblue/test-gen/pull/1988

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 4a2013c3af7ec4db56e8db843889f38ae1c5d40a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77439595

File: jbmc/regression/jbmc-strings/NondetStringBuilderAndBuffer/test.desc [hunk 1]
```diff
@@ -0,0 +1,7 @@
+CORE
+Test.class
+--function Test.testme
+^EXIT=0$
+^SIGNAL=0$
+--
```
Comment (mgudemann) [inline jbmc/regression/jbmc-strings/NondetStringBuilderAndBuffer/test.desc]: test on successful verification?

File: jbmc/src/java_bytecode/java_object_factory.cpp [hunk 2]
```diff
@@ -510,17 +510,22 @@ static codet make_allocate_code(const symbol_exprt &lhs, const exprt &size)
...
-/// Initialize a nondeterministic String structure
-/// \param obj: struct to initialize, must have been declared using
-///   code of the form:
-/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-/// struct java.lang.String { struct \@java.lang.Object;
-///   int length; char *data; } tmp_object_factory$1;
-/// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+/// Check if a structure is a nondeterministic String structure, and if it is
+/// initialize its length and data fields.
+/// \param struct_expr: struct that we may initialize
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/java_object_factory.cpp]: maybe mark as `out` parameter

File: jbmc/src/java_bytecode/java_object_factory.cpp [hunk 3]
```diff
@@ -1042,15 +983,27 @@ void java_object_factoryt::gen_nondet_struct_init(
...
-    exprt zero_object=
+    exprt initial_object =
       zero_initializer(
         struct_type, source_locationt(), ns, nullout);
     irep_idt qualified_clsid = "java::" + id2string(class_identifier);
     set_class_identifier(
-      to_struct_expr(zero_object), ns, symbol_typet(qualified_clsid));
+      to_struct_expr(initial_object), ns, symbol_typet(qualified_clsid));
+
+    // If the initialised type is a special-cased String type, try to set up
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/java_object_factory.cpp]: maybe elaborate more on what special-cased means here

File: jbmc/regression/jbmc-strings/NondetStringBuilderAndBuffer/test.desc [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+type mismatch
```
Comment (allredj) [inline jbmc/regression/jbmc-strings/NondetStringBuilderAndBuffer/test.desc]: I'm not sure I understand what this tests. What was the previous behaviour? The other tests makes sense to me however so I'm still happy.

**Review by allredj — APPROVED** [review]
I'm not extremely familiar with this part of the code, but I can't see anything problematic. And if all the string tests are still passing then I can't really complain! :-)

**Comment by smowton** [comment]
@allredj yep, added a comment. @mgudemann applied your comments.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 8e7b6e7cdf2c66a40f7ba27d527a74f05e635c87).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77490650

**Review by mgudemann — APPROVED** [review]

**Pull request merged**